### PR TITLE
Correcting bug #2568: Don't create intermediate Config classes when building 'octopus' inheritance

### PR DESCRIPTION
**Pull request by Alcolo47**
## Change Summary
Correcting bug #2568.

The problem is coming from the 'inherit_config' that create incrementally intermediate Config class for each bases classes.
This PR create Config class at once with all base Config classes.

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -173,20 +174,36 @@ def prepare_field(cls, field: 'ModelField') -> None:
...
 
-def inherit_config(self_config: 'ConfigType', parent_config: 'ConfigType', **namespace: Any) -> 'ConfigType':
-    if not self_config:
-        base_classes: Tuple['ConfigType', ...] = (parent_config,)
-    elif self_config == parent_config:
-        base_classes = (self_config,)
-    else:
-        base_classes = self_config, parent_config
+def inherit_config(
+    bases: Collection[Type[object]], config_from_namespace: Type[object], config_namespace: Dict[str, Any]
```
Comment (PrettyWood) [inline pydantic/main.py:178]:
```suggestion
    bases: Collection[Type[object]], config_from_namespace: Type[object], config_namespace: Dict[str, Any]
```
```suggestion
    bases: Sequence[Type[object]], config_from_namespace: Type[object], config_namespace: Dict[str, Any]
```

**Review by PrettyWood — COMMENTED** [review]
Thanks!

File: pydantic/main.py (hunk 1 above)
Comment (Alcolo47) [reply pydantic/main.py]: You're right: bases order is the most important.

**Comment by Alcolo47** [comment]
> # [Codecov](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570?src=pr&el=h1) Report
> 
> > Merging [#2570](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570?src=pr&el=desc) ([ccdad01](https://github.com/samuelcolvin/pydantic/commit/ccdad01f97fbaac8374b026569c8b5000fab7567)) into [master](https://codecov.io/gh/samuelcolvin/pydantic/commit/7b7e70557bb80ef5f033e3e9caca68484aae78c5?el=desc) ([7b7e705](https://github.com/samuelcolvin/pydantic/commit/7b7e70557bb80ef5f033e3e9caca68484aae78c5)) will **decrease** coverage by `0.37%`.
> > The diff coverage is `100.00%`.
> 
> > exclamation Current head [ccdad01](https://github.com/samuelcolvin/pydantic/commit/ccdad01f97fbaac8374b026569c8b5000fab7567) differs from pull request most recent head [9d86ceb](https://github.com/samuelcolvin/pydantic/commit/9d86ceb45e3daf9ffd1831f4f3af7d2715428146). Consider uploading reports for the commit [9d86ceb](https://github.com/samuelcolvin/pydantic/commit/9d86ceb45e3daf9ffd1831f4f3af7d2715428146) to get more accurate results
> 
> ```diff
> @@             Coverage Diff             @@
> ##            master    #2570      +/-   ##
> ===========================================
> - Coverage   100.00%   99.62%   -0.38%     
> ===========================================
>   Files           25       25              
>   Lines         5109     5108       -1     
>   Branches      1050     1053       +3     
> ===========================================
> - Hits          5109     5089      -20     
> - Misses           0        4       +4     
> - Partials         0       15      +15     
> ```
> 
> [Impacted Files](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570?src=pr&el=tree) 	Coverage Δ 	
> [pydantic/main.py](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570/diff?src=pr&el=tree#diff-cHlkYW50aWMvbWFpbi5weQ==) 	`100.00% <100.00%> (ø)` 	
> [pydantic/mypy.py](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570/diff?src=pr&el=tree#diff-cHlkYW50aWMvbXlweS5weQ==) 	`95.50% <0.00%> (-4.50%)` 	arrow_down
> [pydantic/env_settings.py](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570/diff?src=pr&el=tree#diff-cHlkYW50aWMvZW52X3NldHRpbmdzLnB5) 	`99.20% <0.00%> (-0.80%)` 	arrow_down
> [pydantic/utils.py](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570/diff?src=pr&el=tree#diff-cHlkYW50aWMvdXRpbHMucHk=) 	`99.31% <0.00%> (-0.69%)` 	arrow_down
> 
> [Continue to review full report at Codecov](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570?src=pr&el=continue).
> 
> > **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> > `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> > Powered by [Codecov](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570?src=pr&el=footer). Last update [7416012...9d86ceb](https://codecov.io/gh/samuelcolvin/pydantic/pull/2570?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

I don't understand the problem:
1. Impacted files are out of my MR
2. on master branch, `make testcov` gives:
```
----------- coverage: platform linux, python 3.9.2-final-0 -----------
Name                             Stmts   Miss Branch BrPart     Cover
---------------------------------------------------------------------
pydantic/__init__.py                15      0      0      0   100.00%
pydantic/_hypothesis_plugin.py     122      0     46      0   100.00%
pydantic/annotated_types.py         12      1      8      1    90.00%
pydantic/class_validators.py       168      0    130      0   100.00%
pydantic/color.py                  190      0     85      0   100.00%
pydantic/dataclasses.py            100      0     46      0   100.00%
pydantic/datetime_parse.py         124      0     64      0   100.00%
pydantic/decorator.py              172      0     72      0   100.00%
pydantic/env_settings.py           126      2     54      0    98.89%
pydantic/error_wrappers.py          88      0     30      0   100.00%
pydantic/errors.py                 278      0      4      0   100.00%
pydantic/fields.py                 514      0    276      0   100.00%
pydantic/generics.py               121      0     69      0   100.00%
pydantic/json.py                    50      1     14      1    96.88%
pydantic/main.py                   522      2    241      0    99.74%
pydantic/mypy.py                   356      1    150      1    99.60%
pydantic/networks.py               271      3     68      1    98.82%
pydantic/parse.py                   36      0     20      0   100.00%
pydantic/schema.py                 421      0    228      0   100.00%
pydantic/tools.py                   29      0      4      0   100.00%
pydantic/types.py                  505      0     96      0   100.00%
pydantic/typing.py                 159     33     68      8    77.53%
pydantic/utils.py                  293      1    144      1    99.54%
pydantic/validators.py             419      0    188      0   100.00%
pydantic/version.py                 17      2      4      0    90.48%
---------------------------------------------------------------------
TOTAL                             5108     46   2109     13    99.04%
```

on this MR branch: 
```
----------- coverage: platform linux, python 3.9.2-final-0 -----------
Name                             Stmts   Miss Branch BrPart     Cover
---------------------------------------------------------------------
pydantic/__init__.py                15      0      0      0   100.00%
pydantic/_hypothesis_plugin.py     122      0     46      0   100.00%
pydantic/annotated_types.py         12      1      8      1    90.00%
pydantic/class_validators.py       168      0    130      0   100.00%
pydantic/color.py                  190      0     85      0   100.00%
pydantic/dataclasses.py            100      0     46      0   100.00%
pydantic/datetime_parse.py         124      0     64      0   100.00%
pydantic/decorator.py              172      0     72      0   100.00%
pydantic/env_settings.py           126      2     54      0    98.89%
pydantic/error_wrappers.py          88      0     30      0   100.00%
pydantic/errors.py                 278      0      4      0   100.00%
pydantic/fields.py                 514      0    276      0   100.00%
pydantic/generics.py               121      0     69      0   100.00%
pydantic/json.py                    50      1     14      1    96.88%
pydantic/main.py                   521      2    247      0    99.74%
pydantic/mypy.py                   356      1    150      1    99.60%
pydantic/networks.py               271      3     68      1    98.82%
pydantic/parse.py                   36      0     20      0   100.00%
pydantic/schema.py                 421      0    228      0   100.00%
pydantic/tools.py                   29      0      4      0   100.00%
pydantic/types.py                  505      0     96      0   100.00%
pydantic/typing.py                 159     33     68      8    77.53%
pydantic/utils.py                  293      1    144      1    99.54%
pydantic/validators.py             419      0    188      0   100.00%
pydantic/version.py                 17      2      4      0    90.48%
---------------------------------------------------------------------
TOTAL                             5107     46   2115     13    99.04%
```
**No coverage lost**

3. I can't access to codecov links (403): I can't investigate deeper.

**Comment by samuelcolvin** [comment]
as mentioned on the issue, I'd like to defer this until v2.

**Comment by samuelcolvin** [comment]
pydantic-core has very [sophisticated logic](https://github.com/samuelcolvin/pydantic-core/blob/5e36e914f1b71c96beb458d69ba965eceaab0010/src/validators/new_class.rs#L164-L201) logic for merging config in sub-models.

Closing this as the approach in V2 will be very different.

**Pull request closed without merging**